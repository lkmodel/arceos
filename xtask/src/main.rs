use clap::Parser;
use serde::Deserialize;
use std::fs::{File, OpenOptions};
use std::io::{Read, Seek, SeekFrom, Write};
use std::path::PathBuf;
use std::process::{Command, Stdio};
use std::str::from_utf8;
use std::{env, fs, io};

const CC: &str = "riscv64-linux-musl-gcc";
const DYNAMIC_FLAG: [&str; 1] = ["-fPIE", ];
const STATIC_FLAG: [&str; 0] = [];

#[derive(Parser, Debug)]
#[command(version, about, long_about = None)]
struct Args {
    /// Architecture to use (e.g., riscv64)
    #[arg(long, default_value = "riscv64")]
    arch: String,

    /// Log level (e.g., warn)
    #[arg(long, short, default_value = "warn")]
    log: String,

    /// Enable QEMU logging (y/n)
    #[arg(long, default_value = "n")]
    qemu_log: String,

    /// App type to run (all, static, dynamic)
    #[arg(long, short)]
    ttype: Option<String>,

    /// Review snapshot and don`t run app
    #[arg(long, short)]
    snapshot: bool,

    /// Which app to run
    app: String,
}

#[derive(Debug, Deserialize)]
struct Config {
    dev: DevConfig,
}

#[derive(Debug, Deserialize)]
struct DevConfig {
    rename: Option<String>,
    ttype: Option<String>,
    dynamic_flags: Option<Vec<String>>,
    static_flags: Option<Vec<String>>,
    snapshot: Option<bool>,
}

fn main() {
    let args = Args::parse();

    if args.snapshot {
        Command::new("cargo")
            .args(&["install", "cargo-insta"])
            .status()
            .expect("Can`t install cargo-insta");
        let current_dir = env::current_dir().expect("Failed to get current directory");
        Command::new("cargo")
            .args(&[
                "insta",
                "review",
                "--workspace-root",
                current_dir.join("payload").join(args.app).to_str().unwrap(),
            ])
            .status()
            .expect("Review failed");
        return;
    }

    let current_dir = env::current_dir().expect("Failed to get current directory");

    // Check and install musl-riscv64
    if !check_installation() {
        if !install_musl_riscv64() {
            eprintln!("Failed to install musl-riscv64");
            return;
        }
    }

    // Checkout the mocklibc branch
    if !check_branch("mocklibc") {
        eprintln!("Failed to switch to mocklibc branch");
        return;
    }

    // Add Rust target
    let status = Command::new("rustup")
        .args(["target", "add", "riscv64gc-unknown-linux-musl"])
        .status()
        .expect("Failed to add Rust target");

    if !status.success() {
        eprintln!("Failed to add Rust target");
        return;
    }

    let is_ci = env::var("CI").is_err();
    if args.app == "all".to_string() || !is_ci {
        traverse_all_app(&current_dir.join("payload"), &|dir: &PathBuf| -> Result<(), String> {
            let arg = Args {
                arch: "riscv64".to_string(),
                log: "warn".to_string(),
                qemu_log: "n".to_string(),
                ttype: None,
                snapshot: false,
                app: dir.iter().last().unwrap().to_str().unwrap().to_string(),
            };
            let config = parse_toml(&current_dir.join("payload").join(&arg.app));
            println!("APP:{}", arg.app);
            dynamic_test(&arg, &current_dir, &config)
        }).expect("Failed to traverse app")
            .iter()
            .for_each(|app| println!("{}\t{}", app[0], app[1]));
        return;
    }

    // Parse the toml file
    let config = parse_toml(&current_dir.join("payload").join(&args.app));
    println!("Config: {:?}", config);
    let mut ttype = config
        .as_ref()
        .and_then(|c| c.dev.ttype.as_ref())
        .unwrap()
        .as_str();
    let args_ttype: &String;
    if args.ttype.is_some() {
        args_ttype = args.ttype.as_ref().unwrap();
        if args_ttype == ttype || ttype == "all" {
            ttype = args_ttype;
        } else {
            eprintln!("Unsupported ttype {}", args_ttype);
            return;
        }
    }

    println!("Architecture: {}", args.arch);
    println!("Log level: {}", args.log);
    println!("QEMU log: {}", args.qemu_log);
    println!("Link type: {}", ttype);
    println!("App: {}", args.app);

    // Build mocklibc
    // let status = Command::new("cargo")
    //     .args([
    //         "build",
    //         "--target",
    //         "riscv64gc-unknown-linux-musl",
    //         "--release",
    //         "-p",
    //         "mocklibc",
    //     ])
    //     .status()
    //     .expect("Failed to build mocklibc");
    //
    // if !status.success() {
    //     eprintln!("Failed to build mocklibc");
    //     return;
    // }

    // Move the built library to the payload directory
    // let lib_path = PathBuf::from("./target/riscv64gc-unknown-linux-musl/release/libmocklibc.so");
    // let payload_path = PathBuf::from(format!("./payload/{}/libmocklibc.so",args.app));
    // std::fs::rename(&lib_path, &payload_path).expect("Failed to move libmocklibc.so");


    // Run tests based on the test type
    match ttype {
        "dynamic" => dynamic_test(&args, &current_dir, &config).unwrap(),
        "static" => static_test(&args, &current_dir, &config).unwrap(),
        "all" => {
            dynamic_test(&args, &current_dir, &config).unwrap();
            println!("----------------------------------------");
            println!(" ");
            println!("----------------------------------------");
            static_test(&args, &current_dir, &config).unwrap();
        }
        _ => eprintln!("Invalid test type"),
    }
}
fn traverse_all_app(path: &PathBuf, cb: &dyn Fn(&PathBuf) -> Result<(), String>) -> io::Result<Vec<Vec<String>>> {
    let mut apps = Vec::new();
    if path.is_dir() {
        for entry in fs::read_dir(path)? {
            let entry = entry?;
            let path = entry.path();
            if path.is_dir() {
                if let Err(e) = cb(&path) {
                    apps.push(vec![path.iter().last().unwrap().to_str().unwrap().to_string(), e]);
                } else {
                    apps.push(vec![path.iter().last().unwrap().to_str().unwrap().to_string(), String::from("OK")]);
                }
            }
        }
    }
    Ok(apps)
}

fn parse_toml(app_path: &PathBuf) -> Option<Config> {
    let toml_path = app_path.join("config.toml");
    if toml_path.exists() {
        let toml_content = fs::read_to_string(toml_path).expect("Failed to read toml file");
        let config: Config = toml::from_str(&toml_content).expect("Failed to parse toml file");
        Some(config)
    } else {
        None
    }
}

fn check_installation() -> bool {
    Command::new("which")
        .arg("riscv64-linux-musl-gcc")
        .output()
        .map(|output| output.status.success())
        .unwrap_or(false)
        || PathBuf::from("/opt/musl_riscv64").exists()
}

fn install_musl_riscv64() -> bool {
    if check_installation() {
        return true;
    }

    if !get_sudo() {
        return false;
    }

    let status = Command::new("git")
        .args(["clone", "https://github.com/richfelker/musl-cross-make.git"])
        .status()
        .expect("Failed to clone musl-cross-make");

    if !status.success() {
        eprintln!("Failed to clone musl-cross-make");
        return false;
    }

    let mut config_mak = std::fs::read_to_string("musl-cross-make/config.mak.dist")
        .expect("Failed to read config.mak.dist");

    config_mak.insert_str(15, "TARGET = riscv64-linux-musl\n");
    config_mak.insert_str(22, "OUTPUT = /opt/musl_riscv64\n");

    std::fs::write("musl-cross-make/config.mak", config_mak).expect("Failed to write config.mak");

    let status = Command::new("make")
        .args(["-j4"])
        .current_dir("musl-cross-make")
        .status()
        .expect("Failed to build musl-cross-make");

    if !status.success() {
        eprintln!("Build failed");
        return false;
    }

    let status = Command::new("sudo")
        .args(["make", "install", "-j4"])
        .current_dir("musl-cross-make")
        .status()
        .expect("Failed to install musl-cross-make");

    if !status.success() {
        eprintln!("Installation failed");
        return false;
    }

    // Add to PATH in .bashrc and .zshrc
    let home_dir = env::var("HOME").expect("Failed to get HOME directory");
    let bashrc_path = PathBuf::from(&home_dir).join(".bashrc");
    let zshrc_path = PathBuf::from(&home_dir).join(".zshrc");

    for rc_file in &[bashrc_path, zshrc_path] {
        if rc_file.exists() {
            let mut rc_content = std::fs::read_to_string(rc_file).expect("Failed to read rc file");
            if !rc_content.contains("/opt/musl_riscv64/bin") {
                rc_content.push_str("\nexport PATH=$PATH:/opt/musl_riscv64/bin\n");
                std::fs::write(rc_file, rc_content).expect("Failed to write rc file");
            }
        }
    }

    println!("Musl RISC-V64 toolchain installation complete");
    std::fs::remove_dir_all("musl-cross-make").expect("Failed to remove musl-cross-make directory");

    // Add musl-riscv64 to PATH
    let mut path = env::var("PATH").unwrap_or_default();
    path.push_str(":/opt/musl_riscv64/bin");
    unsafe {
        env::set_var("PATH", &path);
    }

    true
}

fn get_sudo() -> bool {
    for attempt in 1..=3 {
        println!("Requesting sudo permissions (attempt {}/3)", attempt);
        let status = Command::new("sudo")
            .arg("-v")
            .status()
            .expect("Failed to request sudo permissions");

        if status.success() {
            println!("Sudo permissions granted");
            return true;
        } else {
            println!("Permission denied, please retry");
        }
    }

    eprintln!("Failed to obtain sudo permissions after maximum attempts");
    false
}

fn check_branch(branch_name: &str) -> bool {
    let output = Command::new("git")
        .args(["rev-parse", "--abbrev-ref", "HEAD"])
        .output()
        .expect("Failed to get current branch");

    if output.status.success() {
        let current_branch = String::from_utf8(output.stdout).expect("Failed to parse branch name");
        if current_branch.trim() == branch_name {
            println!("Currently on branch {}", branch_name);
            true
        } else {
            println!("Not on branch {}", branch_name);
            let status = Command::new("git")
                .args(["switch", branch_name])
                .status()
                .expect("Failed to switch branch");

            if status.success() {
                println!("Switched to branch {}", branch_name);
                true
            } else {
                eprintln!("Failed to switch branch");
                false
            }
        }
    } else {
        eprintln!("Failed to get current branch");
        false
    }
}

fn static_test(args: &Args, current_dir: &PathBuf, config: &Option<Config>) -> Result<(), String> {
    let payload_dir = current_dir.join("payload");
    let app_path = payload_dir.join(args.app.as_str());
    build(&app_path, false, config).expect("Failed to build dynamic test");
    run_make(args, current_dir)
}

fn dynamic_test(args: &Args, current_dir: &PathBuf, config: &Option<Config>) -> Result<(), String> {
    let payload_dir = current_dir.join("payload");
    let app_path = payload_dir.join(args.app.as_str());
    build(&app_path, true, config).expect("Failed to build dynamic test");
    run_make(args, current_dir)
}

fn run_make(args: &Args, current_dir: &PathBuf) -> Result<(), String> {
    let status = Command::new("make")
        .args(["defconfig", "ARCH=riscv64"])
        .current_dir(&current_dir)
        .status()
        .expect("Failed to run make defconfig");

    if !status.success() {
        return Err(String::from("make defconfig failed"));
    }

    let status = Command::new("make")
        .args([
            "A=examples/loader",
            &format!("ARCH={}", args.arch),
            &format!("LOG={}", args.log),
            &format!("QEMU_LOG={}", args.qemu_log),
            "run",
        ])
        .current_dir(&current_dir)
        .status()
        .expect("Failed to run make");

    if !status.success() {
        return Err(String::from("make run failed"));
    }
    Ok(())
}

/// Generate a binary file from the ELF file
fn generate_bin(elf_file: &str, working_dir: &PathBuf) -> io::Result<()> {
    let elf_path = working_dir.join(elf_file);
    let bin_path = working_dir.join("apps.bin");

    // Create a 32MB empty file
    let file = File::create(&bin_path)?;
    file.set_len(32 * 1024 * 1024)?; // 32MB

    // Get the size of the ELF file
    let app_size = fs::metadata(&elf_path)?.len();

    // Convert the size to a 16-byte hex string and reverse the byte order
    let size_hex = format!("{:016x}", app_size);
    let mut size_bytes = hex::decode(size_hex).unwrap();
    size_bytes.reverse();

    // Write the size to the binary file (first 8 bytes)
    let mut file = OpenOptions::new().write(true).open(&bin_path)?;
    file.write_all(&size_bytes)?;

    // Write the ELF file content to the binary file (starting at offset 8)
    let mut elf_file = File::open(elf_path)?;
    let mut buffer = Vec::new();
    elf_file.read_to_end(&mut buffer)?;

    file.seek(SeekFrom::Start(8))?;
    file.write_all(&buffer)?;

    // Copy the binary file to the parent directory
    let parent_bin_path = working_dir.parent().unwrap().join("apps.bin");
    fs::rename(&bin_path, &parent_bin_path)?;
    Ok(())
}

/// Build a ELF file
fn build(elf_path: &PathBuf, ttype: bool, config: &Option<Config>) -> io::Result<()> {
    let elf_file = elf_path
        .iter()
        .last()
        .unwrap()
        .to_str()
        .unwrap()
        .split('_')
        .next()
        .unwrap()
        .to_string();
    // Determine the rename value from the config
    let rename = config
        .as_ref()
        .and_then(|c| c.dev.rename.as_ref())
        .unwrap_or(&elf_file)
        .as_str();
    let c_file = format!("{}.c", rename); // C source file
    let elf_output = format!("{}", rename); // Output ELF file

    // Determine the flags based on the config
    let dynamic_flags = config
        .as_ref()
        .and_then(|c| c.dev.dynamic_flags.as_ref())
        .map(|flags| flags.iter().map(|s| s.as_str()).collect::<Vec<_>>())
        .unwrap_or_else(|| DYNAMIC_FLAG.iter().map(|s| *s).collect());

    let static_flags = config
        .as_ref()
        .and_then(|c| c.dev.static_flags.as_ref())
        .map(|flags| flags.iter().map(|s| s.as_str()).collect::<Vec<_>>())
        .unwrap_or_else(|| STATIC_FLAG.iter().map(|s| *s).collect());

    // Compile the C file
    let mut status = Command::new(CC);
    let flags = if ttype {
        dynamic_flags
    } else {
        static_flags
    };
    status
        .args(&["-v", &c_file])
        .args(flags)
        .args(&["-o", &elf_output])
        .current_dir(elf_path)
        .stdout(Stdio::null())
        .stderr(Stdio::null());
    let re = status.status()?;

    if !re.success() {
        return Err(io::Error::new(
            io::ErrorKind::Other,
            "Failed to compile C file",
        ));
    }

    let mut insta_set = insta::Settings::clone_current();
    insta_set.set_snapshot_path(elf_path.join("snapshot"));
    insta_set.set_prepend_module_to_snapshot(false);

    let t_type: String;
    if ttype {
        t_type = "dynamic".to_string();
    } else {
        t_type = "static".to_string();
    }

    // Generate disassembly file
    let output = Command::new("riscv64-linux-musl-objdump")
        .args(&["-d", &elf_output])
        .current_dir(elf_path)
        .output()
        .expect("Failed to run riscv64-linux-musl-objdump");
    let output_file = elf_path.clone();

    if config
        .as_ref()
        .and_then(|c| c.dev.snapshot.as_ref())
        .is_some_and(|b| *b)
    {
        insta_set.set_snapshot_suffix(format!("{}_{}.S", elf_file, t_type));
        insta_set.bind(|| {
            insta::assert_snapshot!(from_utf8(&output.stdout).unwrap());
        });
    }
    fs::write(output_file.join(&format!("{}.S", elf_file)), output.stdout)
        .expect("Failed to write ELF file");

    // Generate ELF info file
    let output = Command::new("riscv64-linux-musl-readelf")
        .args(&["-a", &elf_output])
        .current_dir(elf_path)
        .output()
        .expect("Failed to run riscv64-linux-musl-readelf");
    let output_file = elf_path.clone();
    if config
        .as_ref()
        .and_then(|c| c.dev.snapshot.as_ref())
        .is_some_and(|b| *b)
    {
        insta_set.set_snapshot_suffix(format!("{}_{}.elf", elf_file, t_type));
        insta_set.bind(|| {
            insta::assert_snapshot!(from_utf8(&output.stdout).unwrap());
        });
    }
    fs::write(
        output_file.join(&format!("{}.elf", elf_file)),
        output.stdout,
    )
        .expect("Failed to write ELF file");

    // Generate full disassembly and symbol table
    let output = Command::new("riscv64-linux-musl-objdump")
        .args(&["-x", "-d", &elf_output])
        .current_dir(elf_path)
        .output()
        .expect("Failed to run riscv64-linux-musl-objdump");
    let output_file = elf_path.clone();
    if config
        .as_ref()
        .and_then(|c| c.dev.snapshot.as_ref())
        .is_some_and(|b| *b)
    {
        insta_set.set_snapshot_suffix(format!("{}_{}.dump", elf_file, t_type));
        insta_set.bind(|| {
            insta::assert_snapshot!(from_utf8(&output.stdout).unwrap());
        });
    }
    fs::write(
        output_file.join(&format!("{}.dump", elf_file)),
        output.stdout,
    )
        .expect("Failed to write ELF file");

    // Generate the binary file
    generate_bin(&elf_output, &elf_path)?;

    Ok(())
}

