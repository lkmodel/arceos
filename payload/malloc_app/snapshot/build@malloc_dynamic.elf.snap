---
source: xtask/src/main.rs
description: riscv64-linux-musl-gcc (GCC) 9.4.0
expression: from_utf8(&output.stdout).unwrap()
---
ELF Header:
  Magic:   7f 45 4c 46 02 01 01 00 00 00 00 00 00 00 00 00 
  Class:                             ELF64
  Data:                              2's complement, little endian
  Version:                           1 (current)
  OS/ABI:                            UNIX - System V
  ABI Version:                       0
  Type:                              EXEC (Executable file)
  Machine:                           RISC-V
  Version:                           0x1
  Entry point address:               0x105e0
  Start of program headers:          64 (bytes into file)
  Start of section headers:          10672 (bytes into file)
  Flags:                             0x5, RVC, double-float ABI
  Size of this header:               64 (bytes)
  Size of program headers:           56 (bytes)
  Number of program headers:         7
  Size of section headers:           64 (bytes)
  Number of section headers:         21
  Section header string table index: 20

Section Headers:
  [Nr] Name              Type             Address           Offset
       Size              EntSize          Flags  Link  Info  Align
  [ 0]                   NULL             0000000000000000  00000000
       0000000000000000  0000000000000000           0     0     0
  [ 1] .interp           PROGBITS         00000000000101c8  000001c8
       000000000000001a  0000000000000000   A       0     0     1
  [ 2] .hash             HASH             00000000000101e8  000001e8
       0000000000000044  0000000000000004   A       4     0     8
  [ 3] .gnu.hash         GNU_HASH         0000000000010230  00000230
       0000000000000050  0000000000000000   A       4     0     8
  [ 4] .dynsym           DYNSYM           0000000000010280  00000280
       0000000000000120  0000000000000018   A       5     1     8
  [ 5] .dynstr           STRTAB           00000000000103a0  000003a0
       000000000000005d  0000000000000000   A       0     0     1
  [ 6] .rela.plt         RELA             0000000000010400  00000400
       0000000000000108  0000000000000018  AI       4    14     8
  [ 7] .plt              PROGBITS         0000000000010510  00000510
       00000000000000d0  0000000000000010  AX       0     0     16
  [ 8] .text             PROGBITS         00000000000105e0  000005e0
       0000000000000aa4  0000000000000000  AX       0     0     2
  [ 9] .rodata           PROGBITS         0000000000011088  00001088
       0000000000000490  0000000000000000   A       0     0     8
  [10] .eh_frame         PROGBITS         0000000000011518  00001518
       0000000000000004  0000000000000000   A       0     0     4
  [11] .init_array       INIT_ARRAY       0000000000012e40  00001e40
       0000000000000008  0000000000000008  WA       0     0     8
  [12] .fini_array       FINI_ARRAY       0000000000012e48  00001e48
       0000000000000008  0000000000000008  WA       0     0     8
  [13] .dynamic          DYNAMIC          0000000000012e50  00001e50
       00000000000001b0  0000000000000010  WA       5     0     8
  [14] .got              PROGBITS         0000000000013000  00002000
       0000000000000070  0000000000000008  WA       0     0     8
  [15] .sdata            PROGBITS         0000000000013070  00002070
       0000000000000008  0000000000000000  WA       0     0     8
  [16] .bss              NOBITS           0000000000013078  00002078
       0000000000000038  0000000000000000  WA       0     0     8
  [17] .comment          PROGBITS         0000000000000000  00002078
       0000000000000011  0000000000000001  MS       0     0     1
  [18] .symtab           SYMTAB           0000000000000000  00002090
       0000000000000648  0000000000000018          19    36     8
  [19] .strtab           STRTAB           0000000000000000  000026d8
       000000000000023d  0000000000000000           0     0     1
  [20] .shstrtab         STRTAB           0000000000000000  00002915
       000000000000009a  0000000000000000           0     0     1
Key to Flags:
  W (write), A (alloc), X (execute), M (merge), S (strings), I (info),
  L (link order), O (extra OS processing required), G (group), T (TLS),
  C (compressed), x (unknown), o (OS specific), E (exclude),
  p (processor specific)

There are no section groups in this file.

Program Headers:
  Type           Offset             VirtAddr           PhysAddr
                 FileSiz            MemSiz              Flags  Align
  PHDR           0x0000000000000040 0x0000000000010040 0x0000000000010040
                 0x0000000000000188 0x0000000000000188  R      0x8
  INTERP         0x00000000000001c8 0x00000000000101c8 0x00000000000101c8
                 0x000000000000001a 0x000000000000001a  R      0x1
      [Requesting program interpreter: /lib/ld-musl-riscv64.so.1]
  LOAD           0x0000000000000000 0x0000000000010000 0x0000000000010000
                 0x000000000000151c 0x000000000000151c  R E    0x1000
  LOAD           0x0000000000001e40 0x0000000000012e40 0x0000000000012e40
                 0x0000000000000238 0x0000000000000270  RW     0x1000
  DYNAMIC        0x0000000000001e50 0x0000000000012e50 0x0000000000012e50
                 0x00000000000001b0 0x00000000000001b0  RW     0x8
  GNU_STACK      0x0000000000000000 0x0000000000000000 0x0000000000000000
                 0x0000000000000000 0x0000000000000000  RW     0x10
  GNU_RELRO      0x0000000000001e40 0x0000000000012e40 0x0000000000012e40
                 0x00000000000001c0 0x00000000000001c0  R      0x1

 Section to Segment mapping:
  Segment Sections...
   00     
   01     .interp 
   02     .interp .hash .gnu.hash .dynsym .dynstr .rela.plt .plt .text .rodata .eh_frame 
   03     .init_array .fini_array .dynamic .got .sdata .bss 
   04     .dynamic 
   05     
   06     .init_array .fini_array .dynamic 

Dynamic section at offset 0x1e50 contains 22 entries:
  Tag        Type                         Name/Value
 0x0000000000000001 (NEEDED)             Shared library: [libc.so]
 0x000000000000000c (INIT)               0x10570
 0x000000000000000d (FINI)               0x105a0
 0x0000000000000019 (INIT_ARRAY)         0x12e40
 0x000000000000001b (INIT_ARRAYSZ)       8 (bytes)
 0x000000000000001a (FINI_ARRAY)         0x12e48
 0x000000000000001c (FINI_ARRAYSZ)       8 (bytes)
 0x0000000000000004 (HASH)               0x101e8
 0x000000006ffffef5 (GNU_HASH)           0x10230
 0x0000000000000005 (STRTAB)             0x103a0
 0x0000000000000006 (SYMTAB)             0x10280
 0x000000000000000a (STRSZ)              93 (bytes)
 0x000000000000000b (SYMENT)             24 (bytes)
 0x0000000000000015 (DEBUG)              0x0
 0x0000000000000003 (PLTGOT)             0x13000
 0x0000000000000002 (PLTRELSZ)           264 (bytes)
 0x0000000000000014 (PLTREL)             RELA
 0x0000000000000017 (JMPREL)             0x10400
 0x0000000000000007 (RELA)               0x10400
 0x0000000000000008 (RELASZ)             264 (bytes)
 0x0000000000000009 (RELAENT)            24 (bytes)
 0x0000000000000000 (NULL)               0x0

Relocation section '.rela.plt' at offset 0x400 contains 11 entries:
  Offset          Info           Type           Sym. Value    Sym. Name + Addend
000000013010  000400000005 R_RISCV_JUMP_SLOT 0000000000010530 putchar + 0
000000013018  000500000005 R_RISCV_JUMP_SLOT 0000000000010540 printf + 0
000000013020  000100000005 R_RISCV_JUMP_SLOT 0000000000010550 puts + 0
000000013028  000600000005 R_RISCV_JUMP_SLOT 0000000000010560 malloc + 0
000000013030  000700000005 R_RISCV_JUMP_SLOT 0000000000000000 _init + 0
000000013038  000900000005 R_RISCV_JUMP_SLOT 0000000000010580 calloc + 0
000000013040  000200000005 R_RISCV_JUMP_SLOT 0000000000010590 realloc + 0
000000013048  000a00000005 R_RISCV_JUMP_SLOT 0000000000000000 _fini + 0
000000013050  000b00000005 R_RISCV_JUMP_SLOT 00000000000105b0 __libc_start_main + 0
000000013058  000300000005 R_RISCV_JUMP_SLOT 00000000000105c0 strlen + 0
000000013060  000800000005 R_RISCV_JUMP_SLOT 00000000000105d0 free + 0

The decoding of unwind sections for machine type RISC-V is not currently supported.

Symbol table '.dynsym' contains 12 entries:
   Num:    Value          Size Type    Bind   Vis      Ndx Name
     0: 0000000000000000     0 NOTYPE  LOCAL  DEFAULT  UND 
     1: 0000000000010550     0 FUNC    GLOBAL DEFAULT  UND puts
     2: 0000000000010590     0 FUNC    GLOBAL DEFAULT  UND realloc
     3: 00000000000105c0     0 FUNC    GLOBAL DEFAULT  UND strlen
     4: 0000000000010530     0 FUNC    GLOBAL DEFAULT  UND putchar
     5: 0000000000010540     0 FUNC    GLOBAL DEFAULT  UND printf
     6: 0000000000010560     0 FUNC    GLOBAL DEFAULT  UND malloc
     7: 0000000000000000     0 FUNC    WEAK   DEFAULT  UND _init
     8: 00000000000105d0     0 FUNC    GLOBAL DEFAULT  UND free
     9: 0000000000010580     0 FUNC    GLOBAL DEFAULT  UND calloc
    10: 0000000000000000     0 FUNC    WEAK   DEFAULT  UND _fini
    11: 00000000000105b0     0 FUNC    GLOBAL DEFAULT  UND __libc_start_main

Symbol table '.symtab' contains 67 entries:
   Num:    Value          Size Type    Bind   Vis      Ndx Name
     0: 0000000000000000     0 NOTYPE  LOCAL  DEFAULT  UND 
     1: 00000000000101c8     0 SECTION LOCAL  DEFAULT    1 
     2: 00000000000101e8     0 SECTION LOCAL  DEFAULT    2 
     3: 0000000000010230     0 SECTION LOCAL  DEFAULT    3 
     4: 0000000000010280     0 SECTION LOCAL  DEFAULT    4 
     5: 00000000000103a0     0 SECTION LOCAL  DEFAULT    5 
     6: 0000000000010400     0 SECTION LOCAL  DEFAULT    6 
     7: 0000000000010510     0 SECTION LOCAL  DEFAULT    7 
     8: 00000000000105e0     0 SECTION LOCAL  DEFAULT    8 
     9: 0000000000011088     0 SECTION LOCAL  DEFAULT    9 
    10: 0000000000011518     0 SECTION LOCAL  DEFAULT   10 
    11: 0000000000012e40     0 SECTION LOCAL  DEFAULT   11 
    12: 0000000000012e48     0 SECTION LOCAL  DEFAULT   12 
    13: 0000000000012e50     0 SECTION LOCAL  DEFAULT   13 
    14: 0000000000013000     0 SECTION LOCAL  DEFAULT   14 
    15: 0000000000013070     0 SECTION LOCAL  DEFAULT   15 
    16: 0000000000013078     0 SECTION LOCAL  DEFAULT   16 
    17: 0000000000000000     0 SECTION LOCAL  DEFAULT   17 
    18: 0000000000000000     0 FILE    LOCAL  DEFAULT  ABS crt1.c
    19: 0000000000000000     0 FILE    LOCAL  DEFAULT  ABS crtstuff.c
    20: 0000000000011518     0 OBJECT  LOCAL  DEFAULT   10 
    21: 0000000000010614     0 FUNC    LOCAL  DEFAULT    8 deregister_tm_clones
    22: 0000000000010638     0 FUNC    LOCAL  DEFAULT    8 register_tm_clones
    23: 0000000000010668     0 FUNC    LOCAL  DEFAULT    8 __do_global_dtors_aux
    24: 0000000000013078     1 OBJECT  LOCAL  DEFAULT   16 completed.4702
    25: 0000000000012e48     0 OBJECT  LOCAL  DEFAULT   12 __do_global_dtors_aux_fin
    26: 000000000001069e     0 FUNC    LOCAL  DEFAULT    8 frame_dummy
    27: 0000000000013080    48 OBJECT  LOCAL  DEFAULT   16 object.4707
    28: 0000000000012e40     0 OBJECT  LOCAL  DEFAULT   11 __frame_dummy_init_array_
    29: 0000000000000000     0 FILE    LOCAL  DEFAULT  ABS malloc.c
    30: 0000000000000000     0 FILE    LOCAL  DEFAULT  ABS crtstuff.c
    31: 0000000000011518     0 OBJECT  LOCAL  DEFAULT   10 __FRAME_END__
    32: 0000000000000000     0 FILE    LOCAL  DEFAULT  ABS 
    33: 0000000000012e50     0 OBJECT  LOCAL  DEFAULT   13 _DYNAMIC
    34: 0000000000010510     0 OBJECT  LOCAL  DEFAULT    7 _PROCEDURE_LINKAGE_TABLE_
    35: 0000000000013068     0 OBJECT  LOCAL  DEFAULT   14 _GLOBAL_OFFSET_TABLE_
    36: 0000000000010530     0 FUNC    GLOBAL DEFAULT  UND putchar
    37: 0000000000010f50   308 FUNC    GLOBAL DEFAULT    8 test_calloc_struct
    38: 0000000000010540     0 FUNC    GLOBAL DEFAULT  UND printf
    39: 0000000000013800     0 NOTYPE  GLOBAL HIDDEN   ABS __global_pointer$
    40: 0000000000013070     0 NOTYPE  GLOBAL DEFAULT   15 __SDATA_BEGIN__
    41: 0000000000013000     0 OBJECT  GLOBAL HIDDEN    14 __TMC_END__
    42: 0000000000010550     0 FUNC    GLOBAL DEFAULT  UND puts
    43: 0000000000010560     0 FUNC    GLOBAL DEFAULT  UND malloc
    44: 0000000000013070     0 OBJECT  GLOBAL HIDDEN    15 __dso_handle
    45: 000000000001075a    90 FUNC    GLOBAL DEFAULT    8 test_simple_malloc
    46: 000000000001088c   282 FUNC    GLOBAL DEFAULT    8 test_struct_malloc
    47: 0000000000000000     0 FUNC    WEAK   DEFAULT  UND _init
    48: 0000000000010580     0 FUNC    GLOBAL DEFAULT  UND calloc
    49: 00000000000107b4   216 FUNC    GLOBAL DEFAULT    8 test_array_malloc
    50: 00000000000105e0     0 FUNC    GLOBAL DEFAULT    8 _start
    51: 0000000000010d3c   244 FUNC    GLOBAL DEFAULT    8 test_calloc
    52: 00000000000105f8    28 FUNC    GLOBAL DEFAULT    8 _start_c
    53: 0000000000010590     0 FUNC    GLOBAL DEFAULT  UND realloc
    54: 0000000000010bae   398 FUNC    GLOBAL DEFAULT    8 test_realloc
    55: 00000000000130b0     0 NOTYPE  GLOBAL DEFAULT   16 __BSS_END__
    56: 0000000000013078     0 NOTYPE  GLOBAL DEFAULT   16 __bss_start
    57: 00000000000106c6   148 FUNC    GLOBAL DEFAULT    8 main
    58: 0000000000000000     0 FUNC    WEAK   DEFAULT  UND _fini
    59: 0000000000013000     0 NOTYPE  GLOBAL DEFAULT   14 __DATA_BEGIN__
    60: 0000000000013078     0 NOTYPE  GLOBAL DEFAULT   15 _edata
    61: 00000000000109a6   520 FUNC    GLOBAL DEFAULT    8 test_2d_array_malloc
    62: 00000000000130b0     0 NOTYPE  GLOBAL DEFAULT   16 _end
    63: 0000000000010e30   288 FUNC    GLOBAL DEFAULT    8 test_realloc_string
    64: 00000000000105b0     0 FUNC    GLOBAL DEFAULT  UND __libc_start_main
    65: 00000000000105c0     0 FUNC    GLOBAL DEFAULT  UND strlen
    66: 00000000000105d0     0 FUNC    GLOBAL DEFAULT  UND free

Histogram for bucket list length (total of 3 buckets):
 Length  Number     % of total  Coverage
      0  0          (  0.0%)
      1  0          (  0.0%)      0.0%
      2  1          ( 33.3%)     18.2%
      3  1          ( 33.3%)     45.5%
      4  0          (  0.0%)     45.5%
      5  0          (  0.0%)     45.5%
      6  1          ( 33.3%)    100.0%

Histogram for `.gnu.hash' bucket list length (total of 3 buckets):
 Length  Number     % of total  Coverage
      0  0          (  0.0%)
      1  0          (  0.0%)      0.0%
      2  0          (  0.0%)      0.0%
      3  2          ( 66.7%)     54.5%
      4  0          (  0.0%)     54.5%
      5  1          ( 33.3%)    100.0%

No version information found in this file.
